spec/unit/module_loader/puppetfile_spec.rb

```diff
@@ -0,0 +1,297 @@
+
+      expect { subject.add_module('puppet/test_module', module_opts) }.to raise_error(R10K::Error, /cannot manage content.*is not within/i).and not_change { subject.modules }
+    end
+
+    it 'should disable and not add modules that conflict with the environment' do
+      env = instance_double('R10K::Environment::Base')
+      mod = instance_double('R10K::Module::Base', name: 'conflict', origin: :puppetfile)
+      allow(mod).to receive(:origin=).and_return(nil)
+      loader = described_class.new(basedir: basedir, environment: env)
+      allow(env).to receive(:'module_conflicts?').with(mod).and_return(true)
+
+      allow(R10K::Module).to receive(:new).with('test', anything, anything, anything).and_return(mod)
+      expect { loader.add_module('test', {}) }.not_to change { loader.modules }
+    end
+  end
+
+  describe '#purge_exclusions' do
+    let(:managed_dirs) { ['dir1', 'dir2'] }
+    subject { R10K::ModuleLoader::Puppetfile.new(basedir: '/test/basedir') }
+
+    it 'includes managed_directories' do
+      expect(subject.send(:determine_purge_exclusions, managed_dirs)).to match_array(managed_dirs)
+    end
+
+    context 'when belonging to an environment' do
+      let(:env_contents) { ['env1', 'env2' ] }
+      let(:env) { double(:environment, desired_contents: env_contents) }
+
+      subject { R10K::ModuleLoader::Puppetfile.new(basedir: '/test/basedir', environment: env) }
+
+      it "includes environment's desired_contents" do
+        expect(subject.send(:determine_purge_exclusions, managed_dirs)).to match_array(managed_dirs + env_contents)
+      end
+    end
+  end
+
+  describe '#managed_directories' do
+
+    let(:basedir) { '/test/basedir' }
+    subject { R10K::ModuleLoader::Puppetfile.new(basedir: basedir) }
+
+    before do
+      allow(subject).to receive(:puppetfile_content).and_return('')
+    end
+
+    it 'returns an array of paths that can be purged' do
+      allow(R10K::Module).to receive(:new).with('puppet/test_module', subject.moduledir, hash_including(version: '1.2.3'), anything).and_call_original
+      subject.add_module('puppet/test_module', '1.2.3')
+      subject.load!
+
+      expect(subject.modules.length).to be 1
+      expect(subject.managed_directories).to match_array([subject.moduledir])
+    end
+
+    context "with a module with install_path == ''" do
+      it "basedir isn't in the list of paths to purge" do
+        module_opts = { install_path: '', git: 'git@example.com:puppet/test_module.git' }
+
+        allow(R10K::Module).to receive(:new).with('puppet/test_module', basedir, module_opts, anything).and_call_original
+        subject.add_module('puppet/test_module', module_opts)
+        subject.load!
+
+        expect(subject.modules.length).to be 1
+        expect(subject.managed_directories).to be_empty
+      end
+    end
+  end
+
+  describe 'evaluating a Puppetfile' do
+    def expect_wrapped_error(orig, pf_path, wrapped_error)
```

Comment:
Would it make sense to call this param `error` instead of `orig`? And then maybe `wrapped_error` could be `wrapped_error_type`?